src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -375,6 +376,98 @@ class VariantDatasetFunctions(private val vds: VariantSampleMatrix[Genotype]) ex
     }.copy(vaSignature = finalType)
   }
 
+  def annotateAllelesVDS(other: VariantDataset, code: String, matchStar: Boolean = true): VariantDataset = {
+    import LocusImplicits.orderedKey
+
+    val otherWasSplit = other.wasSplit
+
+    val ec = EvalContext(Map(
+      "global" -> (0, vds.globalSignature),
+      "v" -> (1, TVariant),
+      "va" -> (2, vds.vaSignature),
+      "vds" -> (3, TArray(other.vaSignature)),
+      "aIndices" -> (4, TArray(TInt))
+    ))
+
+    val localGlobalAnnotation = vds.globalAnnotation
+
+    val (paths, types, f) = Parser.parseAnnotationExprs(code, ec, Some(Annotation.VARIANT_HEAD))
+
+    val inserterBuilder = mutable.ArrayBuilder.make[Inserter]
+    val newSignature = (paths, types).zipped.foldLeft(vds.vaSignature) { case (vas, (ids, signature)) =>
+      val (s, i) = vas.insert(signature, ids)
+      inserterBuilder += i
+      s
+    }
+    val inserters = inserterBuilder.result()
+
+    if (vds.wasSplit) {
+      val otherSplit = if (otherWasSplit) other else other.dropSamples().splitMulti(keepStar = matchStar)
+      val (_, aIndexQuerier) = if (otherWasSplit) (null, null) else otherSplit.queryVA("va.aIndex")
+      val newRDD = vds.rdd.orderedLeftJoinDistinct(otherSplit.variantsAndAnnotations)
+        .mapValuesWithKey { case (v, ((va, gs), annotation)) =>
+          val annotations = IndexedSeq(annotation.getOrElse(null))
+          val aIndices = IndexedSeq(annotation.map {
+            ann =>
+              if (otherWasSplit)
+                0
+              else
+                aIndexQuerier(ann).asInstanceOf[Int] - 1
+          }.getOrElse(null))
+
+          ec.setAll(localGlobalAnnotation, v, va, annotations, aIndices)
+          val newVA = f().zip(inserters).foldLeft(va) { case (va, (ann, inserter)) => inserter(va, ann) }
+          (newVA, gs)
+        }.asOrderedRDD
+      vds.copy(rdd = newRDD, vaSignature = newSignature)
+    }
+    else {
+      val otherLocusRDD = other.rdd.mapMonotonic(OrderedKeyFunction(_.locus), { case (v, (va, gs)) => (v, va) })
```

Comment:
if `vds` wasn't split, then this method should just split the right hand side and call `annotateVariantsVDS`. no?